### PR TITLE
`suspicious_double_ref_op`: don't lint on `.borrow()`

### LINKED ISSUE #112489: Confusing new warning suspicious_double_ref_op in beta 1.71.0
### Code

```Rust
use std::borrow::Borrow;

struct S;

trait T: Sized {
    fn foo(self) {}
}

impl T for S {}
impl T for &S {}

fn main() {
    let s = S;
    s.borrow().foo();
    s.foo();
}
```

### Current output

```Shell
Compiling borrow v0.1.0 (/tmp/borrow)
warning: using `.borrow()` on a double reference, which returns `&S` instead of borrowing the inner type
  --> src/main.rs:14:6
   |
14 |     s.borrow().foo();
   |      ^^^^^^^^^
   |
   = note: `#[warn(suspicious_double_ref_op)]` on by default

warning: `borrow` (bin "borrow") generated 1 warning
    Finished dev [unoptimized + debuginfo] target(s) in 0.12s
```

### Desired output

```Shell
Compiling borrow v0.1.0 (/tmp/borrow)
    Finished dev [unoptimized + debuginfo] target(s) in 0.20s
```

### Rationale and extra context

If I have a type `S`, and a trait `T` that is implemented for both `S` and `&S`, `s.borrow()` is a valid way to get a reference `&S` from an object of type `S`. But since beta 1.71, this emits a warning. I have no idea what the warning is trying to convey, what "double reference" is referring to, or what "the inner type" is.

### Other cases

_No response_

### Anything else?

```
$ rustc +beta --version
rustc 1.71.0-beta.3 (78a6ac0a8 2023-06-08)
```

### PR DESCRIPTION
**Pull request by fee1-dead**
closes #112489

### CODE REVIEW

File: compiler/rustc_lint/src/noop_method_call.rs [hunk 1]
```diff
@@ -129,6 +129,9 @@ impl<'tcx> LateLintPass<'tcx> for NoopMethodCall {
                 NoopMethodCallDiag { method: call.ident.name, receiver_ty, label: span },
             );
         } else {
+            if op == "borrow" {
```
Comment (compiler-errors) [inline compiler/rustc_lint/src/noop_method_call.rs:132]: Why not just remove `sym::Borrow` from line 90 above? This seems weirdly hacky.

Comment (Nilstrieb) [inline compiler/rustc_lint/src/noop_method_call.rs:132]: can you add a comment here explaining why borrow makes sense?

Comment (Nilstrieb) [reply compiler/rustc_lint/src/noop_method_call.rs:132]: borrow should still lint for no_op_method_call

File: compiler/rustc_lint/src/noop_method_call.rs [hunk 2]
```diff
@@ -129,6 +129,12 @@ impl<'tcx> LateLintPass<'tcx> for NoopMethodCall {
                 NoopMethodCallDiag { method: call.ident.name, receiver_ty, label: span },
             );
         } else {
+            // FIXME(fee1-dead) stop using stringly typed diagnostic here
+            // If `type_of(x) == T` and `x.borrow()` is used to get `&T`,
+            // gthen that should be allowed
+            if op == "borrow" {
```
Comment (compiler-errors) [inline compiler/rustc_lint/src/noop_method_call.rs:135]: nit:

```suggestion
            if name == sym::noop_method_borrow {
```

**Review by compiler-errors — APPROVED** [review]

**Comment by tspiteri** [comment]
Should this be backported to the 1.71.0 beta? I'm asking because it's listed in the 1.71.0 milestone.

**Comment by cuviper** [comment]
It's too late now for 1.71.0, of course, but it could be considered for 1.71.1. A warn-by-default lint is not such a big deal, but this is a trivial backport (clean cherry-pick), and we're planning to have a point release anyway.

FWIW, I also got a bug for this in CentOS Stream: https://bugzilla.redhat.com/show_bug.cgi?id=2225471

@rustbot label +stable-nominated

**Comment by apiraino** [comment]
Beta backport accepted as per compiler team [on Zulip](https://rust-lang.zulipchat.com/#narrow/stream/238009-t-compiler.2Fmeetings/topic/.5Bweekly.5D.202023-07-27/near/379186994)

@rustbot label +stable-accepted

**Pull request merged**